### ISSUE TITLE
Setting Cache-Control header with serve_static

### ISSUE THREAD

**Issue by ross-byrne** [gleam-wisp/wisp#103]
## Problem
Serving static files using `wisp.serve_static` doesn't give the option of setting a cache-control header. This means static files served this way are not cached by the browser.

I noticed this while serving my static content using the following:
```gleam
use <- wisp.serve_static(req, under: "/static", from: ctx.asset_path)
```

It seems [this code](https://github.com/gleam-wisp/wisp/blob/main/src/wisp.gleam#L1494) is responsible:
```gleam
case simplifile.is_file(path) {
  Ok(True) ->
    response.new(200)
    |> response.set_header("content-type", content_type)
    |> response.set_body(File(path))
  _ -> handler()
}
```

## Possible Solution
I don't have any strong opinions on the best solution. I doubt just throwing in an extra header here is a good idea as it would change the behavior.

Maybe we could add a new function that can take some kind of config?

For context my use case is, I'm serving a Svelte SPA from the backend. It would be nice to have the js files cached without having to re-implement my own version of `wisp.serve_static` :)

Happy to help with implementation if this is something you'd like to add.

Thanks!

**Comment by lpil** [comment]
Thank you! I think this would be great to have, though I'm not entirely sure what the API and behaviour might be. How long would the cache be valid for? What about etags? Should the programmer be able to configure this? etc.

**Comment by ross-byrne** [comment]
> I'm not entirely sure what the API and behaviour might be

Yeah same honestly. My guess is that there would be too much nuance for there to be one good default. I'd like it to cache but others might not depending on the situation. So I'd learn towards having some way to set it.

Maybe making it less about caching specifically and more about being able to set response headers, could be good? I'm not sure if that level of flexibility is useful for others?

**Comment by ross-byrne** [comment]
I get that this could also be a non issue if most people just use a CDN. So I'm happy to just implement my own version of this for now.

**Comment by lpil** [comment]
I'm sure there is a good default out there! Maybe we can look at what other frameworks do and copy them.

While using a CDN is common I think there's a lot of value in supporting non-cloud deployments too, so let's figure that out.

**Comment by ross-byrne** [comment]
OK leave it with me for now. I had a quick look but I think I need to do some more reading on what the different options are.

It seems the two Go based web frameworks I checked, Echo and Chi both leave it up to the developer to set the headers. By just wrapping the handler with another handler to set the response headers.

I looked at Ruby on Rails and they seem to set it for you, but they have a lot of options for caching. So I'd have to dig a bit more to find out what it's doing.

Either way, interesting!

**Comment by ross-byrne** [comment]
@lpil so I've done a bit more reading on the topic and I'm curious about what your take on this is.

The quick summary is: HTML files should either not be cached or have a private, very short lived cache eg. 5 or 10 minutes. 
```http
Cache-Control: max-age:300, private
```

JS and CSS should ideally have versioned names to aid cache invalidation or "cache busting". If versioned, you can set the max-age to a year (1 year is the max a cache can be set). 
```http
Cache-Control: max-age=31536000, immutable
```

Un-versioned files are trickier. Those would need ETags to avoid issues with cached resources going stale. ETags is how Rails handles caching assets with it's asset pipeline but I believe it's off by default. While ETags would be a good feature, the file hashing required would add overhead. So I think that would be a separate thing a user opts into.

### My Thoughts
Caching seems to be very context specific but if we're going to set a default value, the cache should probably be `private` to avoid accidentally leaking sensitive information (see: [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#directives)). We could go with something like: `Cache-Control: public, max-age=604800` which is valid for 7 days. 

My gut feeling is, I'd like to be able to set the response header and/or a config to control how this works. So maybe it's better to leave `serve_static` as it is for now and look at adding an additional function that can take a config? Or something that returns the response so the user can customise it as needed (similar to Go based frameworks). Curious what your thoughts are on this and your appetite for either changing the existing api or adding new functions.

I'll be looking at tackling caching in my own backend soon, maybe next week. So I can prototype some things and share them here if that would be useful? Maybe we can start narrowing down what the API could look like?

Links:  [1](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#caching_static_assets_with_cache_busting) [2](https://simonhearne.com/2022/caching-header-best-practices/#the-solution)

**Comment by lpil** [comment]
That sounds sensible, but it would be good to be able to support the common options people would want. Did you look at what other frameworks do?

**Comment by lpil** [comment]
That's great, thank you Ross. I think it would be good to have this middleware in a new package so we can experiment with APIs and approaches, and then if we are very happy with the design we could potentially move it into `wisp`.

**Comment by ross-byrne** [comment]
Sounds good to me. Any ideas on what the package should be called? I'm not familiar with what an idiomatic name would be for a gleam package. 

I know in the rust world, it's common to have `<package>_mod_<my-thing>`. So we could go with `wisp_mod_caching` or something? That would leave `wisp_caching` free if you wanted to take it over as a separate package instead of upstreaming it straight  into wisp.

**Comment by lpil** [comment]
It's best to avoid using the name of a project as the prefix as it implies it is maintained by the same people, but otherwise not particular convention.

**Comment by ross-byrne** [comment]
Cheers, thanks for the input and guidance here. I don't want to take up too much of your time, so I'll check in again once I have something I've tested and I'm using myself.

**Comment by lpil** [comment]
Love that approach. Extracting libraries from applications is a great way to design code using real-world experience

**Comment by lpil** [comment]
Looks like a good start!

**Comment by ross-byrne** [comment]
> Perhaps for etags it could use a combination of the file mtime and file size?

That is exactly my plan! After doing some digging, building a tag with those two values hashed seems to be the way to go. It's basically how [apache](https://httpd.apache.org/docs/current/mod/core.html#FileETag) handles it. The only difference being apache can also use the INode value but I think that would break if you're using a load balancer.

It seems fairly straight forward to add actually, I'm just trying to get through some work before I can look at it. Once I get that working, I might also share the package in the discord to try get some extra feedback!

**Comment by lpil** [comment]
Wonderful, thank you! A PR sounds great, though we may need to bike-shed on the exact API a little once we have it generally working inside Wisp.

**Comment by ross-byrne** [comment]
Perfect, I can start looking at it this week/weekend.

>  we may need to bike-shed on the exact API a little once we have it generally working inside Wisp

100%. I did my best picking something that looked sane but it definitely needs another pair of eyes.
